arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/raise/Fold.kt

```diff
@@ -224,17 +224,29 @@ public inline fun <Error, A> Raise<Error>.traced(
   @BuilderInference block: Raise<Error>.() -> A,
   trace: (trace: Trace, error: Error) -> Unit
 ): A {
-  val isOuterTraced = this is DefaultRaise && isTraced
-  val nested: DefaultRaise = if (isOuterTraced) this as DefaultRaise else DefaultRaise(true)
+  val nested = DefaultRaise(true)
   return try {
-    block.invoke(nested)
+    block(nested).also { nested.complete() }
   } catch (e: CancellationException) {
+    nested.complete()
     val r: Error = e.raisedOrRethrow(nested)
     trace(Trace(e), r)
-    if (isOuterTraced) throw e else raise(r)
+    // If our outer Raise happens to be traced
+    // Then we want the stack trace to match the inner one
+    try {
+      raise(r)
+    } catch (rethrown: CancellationException) {
+      throw rethrown.withCause(e)
+    }
```

Comment:
```suggestion
    } catch (rethrown: RaiseCancellationException) {
      throw RaiseCancellationException(rethrown.raised, rethrown.raise, e.cause ?: rethrown.cause)
    }
```

Reply:
We can't do this because of #3235 unless we `@PublishedApi` `RaiseCancellationException`.

Reply:
I've been contemplating doing this.. but I've been a bit absent last 2 months due to some personal things keeping me busy.

My rationale for making it public is because I've had a scenario or two where I wanted to differentiate between `RaiseCancellationException`, and (`Job`) `CancellationException`. It's in really low-level edge case scenarios though, but it would allow figuring out if the `CancellationException` comes from Arrow, from KotlinX Coroutines, or from somewhere else.

Perhaps it warrants `@DelicateApi` or something, like KotlinX does to force `@OptIn` so it requires "flagging" low-level code.

WDYT?

Reply:
I think giving users that option is a good idea, especially because we already have tests confirming that `try-catch` can recover from raise. We'd have to make both Raise exceptions public thus. It's another question if we should allow inspecting the error, too.

Reply:
Initially I would say don't expose it if we don't have a need for it.

I got a small WIP, that relies on some new functionality and unofficial support of Kotlin 😅
It adds a `public sealed class`, with expect/actual implementation to disable the stack traces.

I'm going to raise it as a PR targeting this one, so we can discuss further based on a code example ☺️

Reply:
This can now be replaced by the sealed public error for `Raise`, and then this PR is ready to be merged IMO 🥳 🙌